src/main/ruby/truffleruby/core/struct.rb

```diff
@@ -253,12 +253,8 @@ def initialize_copy(other)
   end
 
   def dig(key, *more)
-    result = nil
-    begin
-      result = self[key]
-    rescue IndexError, NameError
-      nil # nothing found with key
-    end
+    return unless _attrs.include?(key.to_sym)
+    result = Primitive.object_hidden_var_get(self, key)
```

Comment:
`Struct#dig` should also handle key being an Integer:
```ruby
ruby -e 'p Struct.new(:a,:b).new(:one, :two).dig(1)' # => 2
```

There is also a partial copy of this logic in `src/main/ruby/truffleruby/core/truffle/diggable.rb`.
To solve that I suggest adding a new private method `dig1` just below `[]` and then essentially copy-paste the logic of `[]` and adapt it in `dig1` (you'll also need a variant/copy of `check_index_var`). And simply call `dig1` in `dig`.

Reply:
`dig1` is meant as "dig with 1 index/dig 1 time", but could also be `read_or_nil` or so to be more explicit.

Reply:
Thanks for feedback!

I rewrote it with `read_or_nil`. I found it hard to make it reused in other places (like `def check_index_var`) so I inlined most of the logic.

I also added a regression test for the edge case you've described.

Reply:
I tried to do a pass to inline `check_index_var` myself, and yes it's hard to factor out because of the exception and their messages. You'd need error codes or something complex like that. Probably best not to.

Reply:
Yes, I was thinking an inlined copy of `check_index_var` and adapting as needed makes sense since there is a lot of differences. Sorry I wasn't clear.